mxfusion/components/distributions/normal.py

```diff
@@ -321,3 +324,328 @@ def _generate_outputs(self, shape):
+    :type precision: Variable
+    :param rand_gen: the random generator (default: MXNetRandomGenerator).
+    :type rand_gen: RandomGenerator
+    :param dtype: the data type for float point numbers.
+    :type dtype: numpy.float32 or numpy.float64
+    :param ctx: the mxnet context (default: None/current context).
+    :type ctx: None or mxnet.cpu or mxnet.gpu
+    """
+    def __init__(self, mean, precision, rand_gen=None, minibatch_ratio=1.,
+                 dtype=None, ctx=None):
+        self.minibatch_ratio = minibatch_ratio
+        if not isinstance(mean, Variable):
+            mean = Variable(value=mean)
+        if not isinstance(precision, Variable):
+            precision = Variable(value=precision)
+
+        inputs = [('mean', mean), ('precision', precision)]
+        input_names = ['mean', 'precision']
+        output_names = ['random_variable']
+        super(MultivariateNormalMeanPrecision, self).__init__(inputs=inputs, outputs=None,
+                                                              input_names=input_names,
+                                                              output_names=output_names,
+                                                              rand_gen=rand_gen, dtype=dtype, ctx=ctx)
+
+    def replicate_self(self, attribute_map=None):
+        """
+        Replicates this Factor, using new inputs, outputs, and a new uuid.
+        Used during model replication to functionally replicate a factor into a new graph.
+
+        :param inputs: new input variables of the factor.
+        :type inputs: a dict of {'name' : Variable} or None
+        :param outputs: new output variables of the factor.
+        :type outputs: a dict of {'name' : Variable} or None
+        """
+        replicant = super(MultivariateNormalMeanPrecision, self).replicate_self(attribute_map)
+        replicant.minibatch_ratio = self.minibatch_ratio
+        return replicant
+
+    @MultivariateNormalMeanPrecisionLogPDFDecorator()
+    def log_pdf(self, mean, precision, random_variable, F=None):
+        """
+        Computes the logarithm of the probability density function (PDF) of the normal distribution.
+
+        :param mean: the mean of the normal distribution.
+        :type mean: MXNet NDArray or MXNet Symbol
+        :param precision: the precision of the distribution.
+        :type precision: MXNet NDArray or MXNet Symbol
+        :param random_variable: the random variable of the normal distribution.
+        :type random_variable: MXNet NDArray or MXNet Symbol
+        :param F: the MXNet computation mode (mxnet.symbol or mxnet.ndarray).
+        :returns: log pdf of the distribution.
+        :rtypes: MXNet NDArray or MXNet Symbol
+        """
+        F = get_default_MXNet_mode() if F is None else F
+        N = mean.shape[-1]
+        c = N * np.log(2 * np.pi)
+        logdetl = -log_determinant(precision)
+        targets = random_variable - mean
+
+        # TODO: Should be a way to do this without loops
```

Comment:
@zhenwendai Ideas here?

I'm not sure what you're computing here, could you write the equation Tom?

Reply:
It's just computing z'Vz where z is the vector z == x - \mu and V is the precision matrix. I couldn't find a broadcasting version of vector-matrix products. It could be done by with sum(diag(z)'V diag(v)), but the current version of diag doesn't yet support specifying the axes (axis1 and axis2) (has been checked in but not in the release yet)